### PR TITLE
#3378 Shabondi sink routes(part 2)

### PR DESCRIPTION
**Pull request by vitojeng**
require by #3378 

Enable data group concept for sink

### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #issue_number)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+import com.island.ohara.common.data.{Row, Serializer}
+import com.island.ohara.kafka.Consumer
+import com.island.ohara.shabondi.Config
+import com.typesafe.scalalogging.Logger
+
+import scala.collection.JavaConverters._
+
+private[sink] class DataGroup(val name: String, brokerProps: String, topicNames: Seq[String], pollTimeout: Duration)
+    extends AutoCloseable {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: please use ohara's Releasable interface

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/WebServer.scala [hunk 2]
```diff
@@ -43,23 +43,22 @@ private[shabondi] class WebServer(val config: Config) extends AbstractWebServer
   import Boot._
   import DefaultDefinitions._
 
+  def start(): Unit = {
+    start(CommonUtils.anyLocalAddress(), config.port, ServerSettings(actorSystem), Some(actorSystem))
+  }
+
   private var _routeHandler: RouteHandler = _
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/WebServer.scala]: consider using ```lazy``` to simplify it. For example
```scala
  private[this] lazy val routeHandler = config.serverType match {
    case SERVER_TYPE_SOURCE => SourceRouteHandler(config)
    case SERVER_TYPE_SINK   => SinkRouteHandler(config)
    case t                  => throw new RuntimeException(s"Invalid server type: $t")
  }

  override def routes: Route = routeHandler.route()
```

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 1 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Cool, thanks for the reminder.

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/WebServer.scala (hunk 2 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/WebServer.scala]: Good point!

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/Config.scala [hunk 3]
```diff
@@ -125,7 +125,7 @@ object DefaultDefinitions {
     .key(SINK_POLL_TIMEOUT)
     .group(coreGroup)
     .orderInGroup(orderNumber)
-    .positiveNumber(500L)
+    .positiveNumber(1500L)
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/Config.scala]: There is a more suitable type - duration. It includes the time unit and the more readable “string” representation. For example, “10 seconds” or “100 days”

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+private[sink] class QueueProducer(
+  val groupName: String,
+  val queue: JQueue[Row],
+  val brokerProps: String,
+  val topicNames: Seq[String],
+  val pollTimeout: Duration
+) extends Runnable
+    with AutoCloseable {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: ditto

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/Config.scala (hunk 3 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/Config.scala]: Ok, thanks!

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      brokerProps
+    )
+    try {
+      while (!stopped.get) {
+        if (!paused.get && queue.isEmpty) {
+          val rows: Seq[Row] = consumer.poll(pollTimeout).asScala.map(_.key.get)
+          rows.foreach(r => queue.add(r))
+          log.trace("    group[{}], queue: {}, rows: {}", groupName, queue.size, rows.size)
+        } else {
+          Thread.sleep(10)
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: TimeUnit.Milliseconds.sleep(10)

The above is more readable

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  override def close(): Unit = {
+    stop()
+  }
+
+  def stop(): Unit = {
+    stopped.set(true)
+  }
+
+  def pause(): Unit = {
+    if (!paused.get) {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Given that both actions are not executed atomically, could you use compareAndSet? For example,
if (paused.compareAndSet(false, true)) {
  do what you want
}

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      paused.set(true)
+      log.info("{} paused.", this.getClass.getSimpleName)
+    }
+  }
+
+  def resume(): Unit = {
+    if (paused.get) {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: ditto

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  }
+}
+
+private[sink] object SinkDataGroups {
+  def apply(config: Config) =
+    new SinkDataGroups(config.brokers, config.sinkFromTopics.map(_.name()), config.sinkPollTimeout)
+}
+
+private[sink] class SinkDataGroups(brokerProps: String, topicNames: Seq[String], pollTimeout: Duration)
+    extends AutoCloseable {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: ditto

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  private val threadPool: ExecutorService =
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Is the resources control a part of “part 3”:)

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 4 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Oops...It's my bad.

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 9 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: yep. In the part 3, there will be a scheduled task to check the idle time of the data group and then release the resources if the idle time up to the maximize idle time.

Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Do you think that the maximize idle time should provide by the user ?

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 8 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: got it!

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 6 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: you are right, thanks!

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 5 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: cool~

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 7 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: got it

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala (hunk 9 above)
Comment (chia7712) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: 至少暴露個參數可以控制，例如預設是無限大，但也可以控制最多幾個group同時存在

Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: Got it.

**Comment by vitojeng** [comment]
not related error

**Pull request merged**